docs/proposals/2018-05-declarative-delete-field.md

```diff
@@ -0,0 +1,83 @@
+# Declarative Delete Field
+
+Concept committed to [during 2018 April 12F2F](https://docs.google.com/document/d/1O7_fws7hwZ6qV3okAjbV5qdFaEq4bt1LYJ9bTu-IwvU)
+
+This document 2018-05.
+
+## Abstract
+
+Add a field to instance and binding to indicate that they have been asked to
+send a DELETE to the backend.
+
+## Motivation
+
+When the Originating-Identity header is used, the broker and refuse a delete
+call when the user info in the Originating-Identity header does not have
+permissions to delete the instance or binding. As this happens after kubernetes
+has accepted the DELETE as an true indication to do a delete, it has set the
+DeletionTimestamp, which is an irreversible step on the way to carrying out the
+necessary actions to delete a kubernetes resource.
+
+## Proposed Design
+
+ - Add a field to indicate deletion
+ - Add a condition to indicate permanent end state of deletion success.
+
+Once the field is activated the controller deprovisions the broker
+resource. When successfully deleted the controller then deletes the
+kubernetes resource.
+
+A feature flag to enable/disable this behavior.
+
+RBAC changes to allow the controller to issue deletes.
+
+Helm chart changes to enable the flag and the rbac rules.
+
+### API Resource Changes
+
+Add a condition for ServiceInstanceConditionType, ServiceBindingConditionType.
+```go
+  ServiceInstanceConditionDeleted ServiceBindingConditionType = "Deleted"
+  ServiceBindingConditionDeleted ServiceBindingConditionType = "Deleted"
+```
+
+Add a field to spec of ServiceInstance, ServiceBinding.
+```go
+  ExistsExternally bool `json:"existsExternally,omitempty"`
+```
+
+This field is set to true by default. The user indicates that it wants
+to delete the backing resource by setting this field to false. The
+controller will see this update and attempt to delete the backing
+broker resource. If the broker delete succeeds, it sets the deleted
+status condition.
+
+If the field is set to false, no updates are allowed besides
+setting the field back to true. 
+
+Once the status is changed to deleted the controller should issue the
+DELETE to the apiserver removing the resource.
+
+### API Server Changes
+
+Additional field results in generated code changes.
+
+Validation/Strategy to prevent changes after a deletion is started and is successful or
+until the deletion is rolled back.
+
+### Controller-Manager Changes
+
+Additional condition to determine delete flow is added. This condition is
```

Comment:
I don't understand the case. If someone calls delete directly after setting that, it's not an update. We can't block deletes anyway, so it'll add the finalizer as usual and go through the existing flow. 

Part of this work should be coming up with rbac rules to prevent access to DELETE. I think this is an extension of rbac that does not currently exist. Created https://github.com/kubernetes-incubator/service-catalog/issues/2061

Reply:
I think you are making the assumption that marking ExistsExternally=false means service catalog will garbage collect that resource at some point in the future, acting as a delayed delete. This is not the usecase we are suggesting at all. We are saying the resource continues to exist in k8s indefinitely until a real DELETE is called, as normal. 

RBAC could be applied to prevent normal people from removing resources and an elevated robot account could garbage collect, but I think this should happen externally to service catalog.

Reply:
I understand the feature you want, and I do not believe it is a feature anyone else wants.

A feature that would do what you want and still be useful to the community would be the described ExistsExternally functionality + a robot script on your side gives you the workflow you want.

Reply:
Who else in the community will use this? So far I've not heard of anyone else who actually wants to adhere to the OSBAPI as strictly as us w.r.t. this aspect, so I'm not that interested in developing a solution for users that don't exist.

Reply:
I don't see this as a OSBAPI feature, I see it more as a service catalog feature.

Reply:
Sure, its a svc-cat feature but its being done in an attempt to be OSBAPI compliant. From past discussions no one else seems to care about this aspect of the spec and are comfortable with the idea of a broker saying "no" after the Kube delete process has started, and can't be stopped. So, I'm looking for a nice UX solution for people who will actually use this feature, not for people who won't.

Reply:
https://github.com/kubernetes-incubator/service-catalog/pull/2037#discussion_r189996298
"at some point in the future" yes, but as instant as it currently exists 
"as a delayed delete" yes, but no more than the existing delay of calling DELETE and waiting for the finalizer to run.
"the resource continues to exist in k8s indefinitely until a real DELETE is called, as normal." no, that does not match my expectations.
https://github.com/kubernetes-incubator/service-catalog/pull/2037#discussion_r190009094
"a robot script on your side gives you the workflow you want." I call that the controller. As part of the existing controller. 

The summary seems to be that we agree that we need a field to indicate that we want this process to start.

Reply:
I don't think we agree at all yet.

Reply:
> I don't see this as a OSBAPI feature, I see it more as a service catalog feature.

@n3wscott it is a confusing feature for Kubernetes users, who are the target users of Service Catalog, not the ones who know OSB API. Normal users won't set the flag, they will just `kubectl delete serviceinstance ...`, they never need to update the Pod spec to shutdown the application, they just delete it.

So I would argue that this delete flag is only useful for emulating OSBAPI deprovisioning semantics to address a corner case of rejected deleted request. If there was no such corner case, users would never ask for deprovisioning without Kubernetes delete.

Reply:
Exactly, a corner case caused by the interaction between the differing behaviors of kubernetes and osbapi.